_posts/2024-04-15-heapdump-oome-native.adoc

```diff
@@ -0,0 +1,258 @@
+---
+layout: post
+title: 'Obtaining heap dump on OutOfMemoryError with Quarkus native'
+date: 2024-04-15
+tags: native heapdump oome epsilongc jol
+synopsis: 'Generating and inspecting heap dumps in Quarkus native when `OutOfMemoryError` errors occur'
+author: galderz
+---
+
+Starting with GraalVM for JDK 21,
+native executables can run with the `-XX:+HeapDumpOnOutOfMemoryError` option to generate a heap dump when a `java.lang.OutOfMemoryError` is thrown.
+In this blog post we will explore how to use the flag, we will inspect what a native image heap dump looks like and how it compares with one produced by HotSpot.
+
+To see this flag in action, we need to manufacture a situation where a Quarkus application runs out of memory.
+One easy way to achieve this is to configure the application with a garbage collector that doesn't do any memory reclamation,
+and that's exactly what the Epsilon GC does.
+Once the Quarkus application is running with Epsilon GC,
+apply some load and within a short space of time it will run out of memory and produce a heap dump.
+
+Let’s start by producing a heap dump with a Quarkus native executable.
+A good starting point would be a Quarkus application that simply responds to an HTTP endpoint request, e.g. the
+https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
+Once checked out,
+build a Quarkus native executable with GraalVM for JDK 21 configuring Epsilon GC.
+For Quarkus native mode, selecting the GC needs to be done at build time:
+
+[source,bash]
+----
+$ ./mvnw package -DskipTests -Dnative -Dquarkus.native.additional-build-args=--gc=epsilon
+----
+
```

Comment:
I think this can be shortened a bit:

```suggestion
Let’s do this using a Quarkus application that simply responds to an HTTP endpoint request as a starting point, e.g. the
https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
Once checked out,
we build a Quarkus native executable with GraalVM for JDK 21 configuring it to use Epsilon GC:

[source,bash]
----
$ ./mvnw package -DskipTests -Dnative -Dquarkus.native.additional-build-args=--gc=epsilon
----

Note: The GC selection needs to be done at build time for Quarkus native applications.

```